lib/gemcutter/middleware/admin_auth.rb

```diff
@@ -0,0 +1,141 @@
+require_relative "../middleware"
+
+module GitHubOAuthable
+  extend ActiveSupport::Concern
+
+  INFO_QUERY = <<~GRAPHQL.freeze
+    query($organization_name:String!) {
+      viewer {
+        name
+        login
+        email
+        avatarUrl
+        id
+        organization(login: $organization_name) {
+          login
+          name
+          viewerIsAMember
+          teams(first: 20, role: MEMBER) {
+            edges {
+              node {
+                name
+                slug
+              }
+            }
+          }
+        }
+      }
+    }
+  GRAPHQL
+
+  included do
+    def admin_user
+      return unless (cookie = cookies.encrypted[admin_cookie_name].presence)
+      Admin::GitHubUser.admins.find(cookie)
+    end
+
+    def admin_logout
+      cookies.delete(admin_cookie_name)
+      redirect_to root_path
+    end
+
+    def admin_github_login!(token:)
+      info_data = fetch_admin_user_info(token)
+      user = Admin::GitHubUser.find_or_initialize_by(github_id: info_data.dig(:viewer, :id))
+      user.is_admin = true # will be set to false if the is_admin validation fails
+      user.oauth_token = token
+      user.info_data = info_data
+      if user.invalid? && user.errors.group_by_attribute.keys == %i[is_admin]
+        is_admin_error = ActiveModel::ValidationError.new(user)
+        user.is_admin = false
+      end
+      user.save!
```

Comment:
is it possible for us to skip saving any information about GitHub users that aren't admins? or do we need that data for some action later?

Reply:
> is it possible for us to skip saving any information about GitHub users that aren't admins? or do we need that data for some action later?

Yeah it probably is in the case it's a new user, but we need to persist the change for an existing user so we know they're no longer an admin

Reply:
Oh yeah, that makes sense. I also realized I was worried about DDoS but that doesn't make sense because the DoS is limited by having valid GitHub accounts and going through the oauth dance, so... I think this is probably fine.